src/toxcore/dht/server_ext.rs

```diff
@@ -34,7 +35,12 @@ impl ServerExt for Server {
                 Err(ref e) => {
                     error!("packet receive error = {:?}", e);
                     // ignore packet decode errors
-                    e.as_fail().downcast_ref::<DecodeError>().is_none()
+                    let fail: &Fail = e;
+                    if let Some(cause) = fail.cause() {
+                        cause.downcast_ref::<DecodeError>().is_none()
```

Comment:
You shouldn't need downcast here anymore, you can just check the error kind. The logic behind this check is that if it's packet decode error we continue the processing, but if it's `IoError` thrown by tokio we terminate the processing.

Reply:
@kurnevsky how would you rewrite it?

Reply:
@kpp first your previous remarks should be fixed. Then we can just check if the kind is whatever  will be instead of `EncodeErrorKind::String`. I can show later if @NamsooCho has troubles.